cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java

```diff
@@ -205,6 +214,76 @@ public ConnectorDag(Collection<Connection> connections, Set<String> reduceNodes,
     return addedAlready;
   }
 
+  /**
+   * Isolate the specified node by inserting a connector in front of and behind the node.
+   * If all inputs into the the node are sources, a connector will not be inserted in front.
+   * If all outputs from the node are sinks, a connector will not be inserted after.
+   * Other connectors count as both a source and a sink.
+   */
+  private void isolate(String node, Set<String> addedAlready) {
+    if (!nodes.contains(node)) {
+      throw new IllegalArgumentException(String.format("Cannot isolate node %s because it is not in the dag.", node));
+    }
+
+    /*
+         If an input into this node is not a source and not a connector,
+         or if the input has another output besides this node, insert a connector in front of this node.
+
+         For example, if we're isolating n2, we need to insert a connector in front if we have a dag like:
+
+              |--> n2
+         n1 --|
+              |--> n3
+
+         but not if our dag looks like:
+
+         n1 --> n2 --> n3
+     */
+    boolean shouldInsert = false;
+    for (String input : incomingConnections.get(node)) {
+      if (connectors.contains(input)) {
+        continue;
+      }
+      if (outgoingConnections.get(input).size() > 1 || !sources.contains(input)) {
```

Comment:
Wouldn't we still want to isolate it, even if its stage right after a source?
For instance, if its a source that isn't supported in spark, isolating it will enable running this phase in spark, with the actual source running in mr.